### PR TITLE
Adds tools.cli command parsing and improved option behavior for serialization commands

### LINKED ISSUE #30366: Serdes v2: make commands consistent
Serdes v2 should have the following flags on `export`:

Options:
           -c, --collection ID             Export only specified ID; may occur multiple times
           -C, --no-collections            Do not export any collections or contents; not compatible with -c
           -D, --no-data-model             Do not export any data model entities; useful for subsequent exports
           -f, --include-field-values      Include field values along with field metadata
           -s, --include-database-secrets  Include database connection details (in plain text; use caution)

### LINKED ISSUE #30433: Serdes v2: add again the deprecation warning
Revert the revert:
https://github.com/metabase/metabase/pull/29529

### PR DESCRIPTION
**Pull request by john-metabase**
This PR is focused on reworking `import` and `export` command option handling, but includes a few big (bonus) changes to the `help` command output thanks to the new data-driven argument parsing that is available through the `:arg-spec` metadata functionality in `metabase.cmd`.

To illustrate the changes, here is the old `help` output:

```
export [path & options]
         This command is in development. For now, use [[dump]]. Serialize a Metabase into directory `path`. Options: --collections [collection-id-list] - a comma-separated list of IDs of collection to export --include-field-values - flag, default false, controls export of field values --include-database-secrets - flag, default false, include database connection details
```

New `help` (specifically `help export`) output:

```
export path & options
         Serialize Metabase instance into directory at `path`.
         Options:
           -u, --user EMAIL                Include collections owned by the specified user
           -c, --collection ID             Export only specified ID; may occur multiple times.
               --collections ID_LIST       (Legacy-style) Export collections in comma-separated list of IDs, e.g. '123,456'.
           -C, --no-collections            Do not export any collections or contents; overrides -c.
           -D, --no-data-model             Do not export any data model entities; useful for subsequent exports.
           -f, --include-field-values      Include field values along with field metadata.
           -s, --include-database-secrets  Include database connection details (in plain text; use caution).
```

Some of the bonus features:

* help with no args to list all commands, or one specific command
* unrecognized commands print a list of allowed commands
* unrecognized args (in commands that use arg-spec) are identified and help text for the command is printed

Resolves #30366. Resolves #30433.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30435)

### CODE REVIEW

File: enterprise/backend/src/metabase_enterprise/serialization/cmd.clj [hunk 1]
```diff
@@ -186,11 +186,7 @@
   (let [opts (cond-> opts
                collections
                (assoc :targets (for [c collections] ["Collection" c])))]
-    ;; if we have `:targets` (either because we created them from `:collections`, or because they were specified
-    ;; elsewhere) use [[v2.extract/extract-subtrees]]
-    (if (:targets opts)
-      (v2.extract/extract-subtrees opts)
-      (v2.extract/extract-metabase opts))))
+    (v2.extract/extract opts)))
```
Comment (john-metabase) [inline enterprise/backend/src/metabase_enterprise/serialization/cmd.clj:189]: Moving this logic to the `extract` namespace rather than the `cmd` one.

**Review by bshepherdson — DISMISSED** [review]
> Reviewed 4 of 11 files at r1, 7 of 8 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30435)* status: all files reviewed, 2 unresolved discussions (waiting on @john-metabase and @luizarakaki)

___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 32 at r5](https://reviewable.io/reviews/metabase/metabase/30435#-NUYOpF-00EmoM3BVAhZ:-NUYOpF-00EmoM3BVAh_:b-qgbpor) ([raw file](https://github.com/metabase/metabase/blob/6fe536f6b50f341883ad2ef4d2070d55af952eb7/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L32)):*
> ```Clojure
> 
>     (empty? targets)
>     (into (map first targets))
> ```

Backwards logic? shouldn't the targets be pulled in if not empty?



_Code quote:_
```Clojure
    (empty? targets)
    (into (map first targets))
```

**Review by john-metabase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30435)* status: 12 of 13 files reviewed, 2 unresolved discussions (waiting on @bshepherdson and @luizarakaki)

___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 32 at r5](https://reviewable.io/reviews/metabase/metabase/30435#-NUYOpF-00EmoM3BVAhZ:-NUbCJ3ADEb3W6FE1ceN:by73yry) ([raw file](https://github.com/metabase/metabase/blob/6fe536f6b50f341883ad2ef4d2070d55af952eb7/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L32)):*
<details><summary><i>Previously, bshepherdson (Braden Shepherdson) wrote…</i></summary><blockquote>

Backwards logic? shouldn't the targets be pulled in if not empty?
</blockquote></details>

When `targets` is not empty, we include all content models, so this code was not just wrong but totally unnecessary. I removed it and added a comment to make sense of how it actually works.

**Review by metamben — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30435)* status: 12 of 13 files reviewed, 12 unresolved discussions (waiting on @bshepherdson, @john-metabase, and @luizarakaki)

___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 28 at r6](https://reviewable.io/reviews/metabase/metabase/30435#-NUgDLTO10PZsFP26APL:-NUgDLTO10PZsFP26APM:undefined) ([raw file](https://github.com/metabase/metabase/blob/e16e9b00dfc338e4f5ab045117b21f379ef57e3c/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L28)):*
> ```Clojure
>     (conj "FieldValues")
> 
>     ;; If `targets` is not specifieid, or if it is a non-empty collection, then we
> ```





_Suggestion:_
```Clojure
specified
```
___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 29 at r6](https://reviewable.io/reviews/metabase/metabase/30435#-NUgEQK5CruDzcBZ0LRL:-NUgEQK5CruDzcBZ0LRM:undefined) ([raw file](https://github.com/metabase/metabase/blob/e16e9b00dfc338e4f5ab045117b21f379ef57e3c/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L29)):*
> ```Clojure
> 
>     ;; If `targets` is not specifieid, or if it is a non-empty collection, then we
>     ;; extract all content models. If `targets` is an empty, collection, we do not export
> ```


___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 40 at r6](https://reviewable.io/reviews/metabase/metabase/30435#-NUgElD5-5WYq7ienauJ:-NUgElD5-5WYq7ienauK:bripw8y) ([raw file](https://github.com/metabase/metabase/blob/e16e9b00dfc338e4f5ab045117b21f379ef57e3c/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L40)):*
> ```Clojure
>   "Returns target seq filtered on given model name"
>   [targets model-name]
>   (filter #(= model-name (first %)) targets))
> ```

Cam me told once this Yoda speak was.



_Code quote:_
```Clojure
(= model-name (first %))
```
___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 56 at r6](https://reviewable.io/reviews/metabase/metabase/30435#-NUgFkz22WvsZEZ2hpx_:-NUgFkz22WvsZEZ2hpxa:undefined) ([raw file](https://github.com/metabase/metabase/blob/e16e9b00dfc338e4f5ab045117b21f379ef57e3c/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L56)):*
> ```Clojure
>     ;; start with the special "nil" root collection ID
>     (-> #{nil}
>         (into (map :id roots))
> ```





_Suggestion:_
```Clojure
(map :id) roots
```
___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 57 at r6](https://reviewable.io/reviews/metabase/metabase/30435#-NUgFtRR1frEOx6vGYB4:-NUgFtRR1frEOx6vGYB5:undefined) ([raw file](https://github.com/metabase/metabase/blob/e16e9b00dfc338e4f5ab045117b21f379ef57e3c/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L57)):*
> ```Clojure
>     (-> #{nil}
>         (into (map :id roots))
>         (into (mapcat collection/descendant-ids roots)))))
> ```





_Suggestion:_
```Clojure
(mapcat collection/descendant-ids) roots
```
___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 66 at r6](https://reviewable.io/reviews/metabase/metabase/30435#-NUgNgcD5HP7iHMS_9no:-NUgNgcD5HP7iHMS_9np:b-w2unv2) ([raw file](https://github.com/metabase/metabase/blob/e16e9b00dfc338e4f5ab045117b21f379ef57e3c/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L66)):*
> ```Clojure
>   (let [extract-opts (assoc opts :collection-set (collection-set-for-user user-id))]
>     (eduction cat (for [model (model-set opts)]
>                     (serdes/extract-all model extract-opts)))))
> ```

I don't quite understand why we need such a mix of list comprehension and eduction. Can't we just stick to one of them? (Probably to eduction, if performance matters.)



_Code quote:_
```Clojure
    (eduction cat (for [model (model-set opts)]
                    (serdes/extract-all model extract-opts)))))
```
___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 88 at r6](https://reviewable.io/reviews/metabase/metabase/30435#-NUgJXJX8V5BU3GDVvrD:-NUgJXJX8V5BU3GDVvrE:undefined) ([raw file](https://github.com/metabase/metabase/blob/e16e9b00dfc338e4f5ab045117b21f379ef57e3c/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L88)):*
> ```Clojure
>   Returns nil if there are no escaped values, which is useful for a test."
>   [targets]
>   (let [collection-ids (set (map second (targets-of-type targets "Collection")))
> ```





_Suggestion:_
```Clojure
(into #{} (map second) (targets-of-type targets "Collection")
```
___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 92 at r6](https://reviewable.io/reviews/metabase/metabase/30435#-NUgJnX3CsCURlkz0Jnd:-NUgJnX3CsCURlkz0Jne:undefined) ([raw file](https://github.com/metabase/metabase/blob/e16e9b00dfc338e4f5ab045117b21f379ef57e3c/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L92)):*
> ```Clojure
>                             (mapcat metabase.models.collection/descendant-ids)
>                             set
>                             (set/union (set collection-ids)))
> ```





_Suggestion:_
```Clojure
        collection-set (into collection-ids
                             (mapcat metabase.models.collection/descendant-ids)
                             (t2/select Collection :id [:in collection-ids])))
```
___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 191 at r6](https://reviewable.io/reviews/metabase/metabase/30435#-NUgLyBi066DrqtsPmE7:-NUgLyBi066DrqtsPmE8:b30k8mo) ([raw file](https://github.com/metabase/metabase/blob/e16e9b00dfc338e4f5ab045117b21f379ef57e3c/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L191)):*
> ```Clojure
>                         (filter #(export? (first %)))
>                         (group-by first)
>                         (m/map-vals #(set (map second %))))]
> ```

We use `update-vals` nowadays.



_Code quote:_
```Clojure
m/map-vals
```
___
*[`enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj` line 194 at r6](https://reviewable.io/reviews/metabase/metabase/30435#-NUgN6Zz5w3lpM0xf2pK:-NUgN6Zz5w3lpM0xf2pL:b-w2unv2) ([raw file](https://github.com/metabase/metabase/blob/e16e9b00dfc338e4f5ab045117b21f379ef57e3c/enterprise/backend/src/metabase_enterprise/serialization/v2/extract.clj#L194)):*
> ```Clojure
>       (eduction cat (for [[model ids] by-model]
>                       (eduction (map #(serdes/extract-one model opts %))
>                                 (db/select-reducible (symbol model) :id [:in ids])))))))
> ```

I don't quite understand why we need such a mix of list comprehension and eduction. Can't we just stick to one of them? (Probably to eduction, if performance matters.)



_Code quote:_
```Clojure
      (eduction cat (for [[model ids] by-model]
                      (eduction (map #(serdes/extract-one model opts %))
                                (db/select-reducible (symbol model) :id [:in ids])))))))
```

**Review by metamben — APPROVED** [review]
> Reviewed 2 of 2 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/30435)* status: all files reviewed, 2 unresolved discussions (waiting on @bshepherdson and @luizarakaki)

**Pull request merged**